feat: 4.9 article (#1106)
withastro/astro.build: src/content/blog/astro-490.mdx

---
title: "Astro 4.9"

> description: "Astro 4.8 is now available! This release includes the long-awaited Container API, stabilized experimental features, and more."

Reviewer: ```suggestion
description: "Astro 4.9 is now available! This release includes the long-awaited Container API, stabilized experimental features, and more."
```
Sharing a preview of the post with Joe and seeing the description in the preview card, FTW!  😅

- [**Experimental: Container API**](#experimental-container-api)
- [**Stabilized experimental features**](#stabilized-experimental-features)

To upgrade an existing project, use the automated `@astrojs/upgrade` CLI tool. Alternatively, upgrade manually by running the upgrade command for your package manager:

```sh
# Recommended:
npx @astrojs/upgrade

# Manual:
npm install astro@latest
pnpm upgrade astro --latest
yarn upgrade astro --latest
```

## Experimental: Container API

Astro 4.9 introduces a first look at the long-awaited Container API. This API allows you to render Astro components outside of an Astro application. This is similar to how you can render React components on the server using libraries like `react-dom/server`, or Preact components using `preact-render-to-string`.

Here's an example of how you can use the Container API to render an Astro component to a string:

```ts
import { experimental_AstroContainer as AstroContainer } from 'astro/container';
import Component from './components/Component.astro';

const container = await AstroContainer.create();
console.log(await container.renderToString(Component, { props: {}, slots: {} }))
```

> The posibilities that this API adds are endless, but this is just a first step. One of our main focus at this stage is to make it easier to render Astro components inside test frameworks powered by Vite, such as Vitest.

Reviewer: ```suggestion
The posibilities that this API adds are endless, but this is just a first step. Our main focus at this stage is to make it easier to render Astro components inside test frameworks powered by Vite, such as Vitest.
```
Or "one of our main focuses" if you like, but I think it's true that we are considering this THE main focus right now!

```ts
import { experimental_AstroContainer as AstroContainer } from 'astro/container';
import { expect, test } from 'vitest';
import Card from '../src/components/Card.astro';

test('Card with slots', async () => {
	const container = await AstroContainer.create();
	const result = await container.renderToString(Card, {
		slots: {
			default: 'Card content',
		},
	});

	expect(result).toContain('This is a card');
	expect(result).toContain('Card content');
});
```

> To get started with testing Astro components with Vitest, we made [a new example](https://github.com/withastro/astro/tree/main/examples/container-with-vitest) you can checkout, or create a new project with directly:

Reviewer: ```suggestion
To get started testing Astro components with Vitest, we made [a new example](https://github.com/withastro/astro/tree/main/examples/container-with-vitest) you can checkout, or create a new project with directly:
```

```shell
npm create astro@latest -- --template container-with-vitest
```

[Our testing guide](https://docs.astro.build/en/guides/testing/#vitest-and-container-api) has also been updated with more information on how to use the Container API for testing. Test everything!

> To learn more about the Container API and how to use it, check out the [Container API documentation](https://docs.astro.build/en/reference/container-reference/). To give feedback on this feature, visit the [Container API RFC](https://github.com/withastro/roadmap/pull/916).

Reviewer: ```suggestion
To learn more about the Container API and how to use it, check out the [Container API documentation](https://docs.astro.build/en/reference/container-reference/). To give feedback on this experimental and developing feature, visit the [Container API RFC](https://github.com/withastro/roadmap/pull/916).
```
or just "experimental", but felt like this was as good place to add info like "stuff is happening!" and reemphasize experimental.

Author: That's fine with me!

## Stabilized experimental features

> This release stabilize the CRSF protection (introduced in Astro 4.6) and i18n domains support (introduced in Astro 4.3) features. As such, it is no longer necessary to enable the `experimental.csrfProtection` and `experimental.i18nDomains` options in your Astro config to use these features.

Reviewer: ```suggestion
This release stabilize the CRSF protection (introduced in [Astro 4.6](/blog/astro-460/#experimental-support-for-csrf-protection)) and i18n domains support (introduced in [Astro 4.3](/blog/astro-430/#experimental-add-domain-support-for-i18n)) features. It is no longer necessary to enable the `experimental.csrfProtection` and `experimental.i18nDomains` options in your Astro config to use these features.
```
Maybe a good chance to link back to their original blog posts, so people can read more about them as they were described then?

Author: I wanted to, but was worried it would cause people to click away from the article too much. Fine with me if you want to do it, though! I trust you!

> The `experimental.security.csrfProtection` option has been removed in favor of a new `security.checkOrigin` option. To enable CRSF protection, set `security.checkOrigin` to `true` in your Astro config and remove the `experimental.security.csrfProtection` option if you have it enabled:

Reviewer: ```suggestion
The `experimental.security.csrfProtection` option has been removed in favor of a new `security.checkOrigin` option. To enable cross-site request forgery protection, set `security.checkOrigin` to `true` in your Astro config and remove the `experimental.security.csrfProtection` option if you have it enabled:
```
I don't think we've written this out fully yet, and since there is so much `csrf` already in the option names here, seems like a good place to expand.

```diff
import { defineConfig } from "astro/config";

export default defineConfig({
+	security: {
+    checkOrigin: true
+  }
-	experimental: {
-    security: {
-      csrfProtection: {
-        origin: true
-      }
-    }
-	},
});
```

As for i18n domains, you can keep your configuration as is, just without the `experimental.i18nDomains` option:

```diff
import { defineConfig } from "astro/config";

export default defineConfig({
  site: "https://example.com",
  output: "server",
  adapter: node({
    mode: 'standalone',
  }),
  i18n: {
    defaultLocale: "en",
    locales: ["en", "fr", "pt-br", "es"],
    prefixDefaultLocale: false,
    domains: {
      fr: "https://fr.example.com",
      es: "https://example.es",
    },
  },
-  experimental: {
-    i18nDomains: true,
-  },
});
```

For more information on these features, see the [security](https://docs.astro.build/en/reference/configuration-reference/#security) and [i18n domains](https://docs.astro.build/en/guides/internationalization/#domains) sections of the Astro documentation respectively.

## Bug Fixes

> As it do be, Astro 4.9 includes more bug fixes and smaller improvements that couldn't make it into this post! Check out the full [release notes](https://github.com/withastro/astro/blob/refs/heads/main/packages/astro/CHANGELOG.md#490) to learn more!

Reviewer: I simply cannot change this.

Author: It is very important that people know there's a human changing the text there on every update.

</BlogSummary>

Full release highlights include:

- [**Experimental: Container API**](#experimental-container-api)
- [**React 19 support for Astro Actions**](#react-19-support)
- [**Stabilized experimental features**](#stabilized-experimental-features)

To upgrade an existing project, use the automated `@astrojs/upgrade` CLI tool. Alternatively, upgrade manually by running the upgrade command for your package manager:

```sh
# Recommended:
npx @astrojs/upgrade

# Manual:
npm install astro@latest
pnpm upgrade astro --latest
yarn upgrade astro --latest
```

## Experimental: Container API

Astro 4.9 introduces a first look at the long-awaited Container API. This API allows you to render Astro components outside of an Astro application. This is similar to how you can render React components on the server using libraries like `react-dom/server`, or Preact components using `preact-render-to-string`.

Here's an example of how you can use the Container API to render an Astro component to a string:

```ts
import { experimental_AstroContainer as AstroContainer } from 'astro/container';
import Component from './components/Component.astro';

const container = await AstroContainer.create();

> console.log(await container.renderToString(Component, { props: {}, slots: {} }))

Reviewer 2: Maybe for this initial example of the API it’s worth simplifying it to just show “hey you can render `Component` to a string”:

```suggestion
console.log(await container.renderToString(Component));
```

To use React 19 in your Astro project, [follow the React 19 beta installation guide.](https://react.dev/blog/2024/04/25/react-19-upgrade-guide#installing) This unlocks React's new form actions feature, which lets you call Actions from a `<form>` and track "pending" and "result" values in your components.

This example uses an Astro Action called `like()` that accepts a `postId` and returns the number of likes. With React 19, you can pass this action to the `useActionState()` hook to track the result. Apply the `experimental_withState()` function to apply progressive enhancement information automatically:

```tsx
// src/components/Like.tsx
import { actions } from 'astro:actions';
import { useActionState } from 'react';
import { experimental_withState } from '@astrojs/react/actions';

export function Like({ postId }: { postId: string }) {
	const [state, action, pending] = useActionState(
		experimental_withState(actions.like),
		0, // initial likes
	);

	return (
		<form action={action}>
			<input type="hidden" name="postId" value={postId} />
      <button disabled={pending}>{state} ❤️</button>
		</form>
	);
}
```

> You can also access the state stored by `useActionState()` from your action `handler` using `experimental_getActionState()`. [See the CHANGELOG](TODO:link) for usage examples.

Reviewer: ```suggestion
You can also access the state stored by `useActionState()` from your action `handler` using `experimental_getActionState()`. [See the CHANGELOG](https://github.com/withastro/astro/blob/refs/heads/main/packages/astro/CHANGELOG.md#490) for usage examples.
```

If you don't need to track the state, you can also apply Astro Actions directly to the `action` property on any React `<form>`. This will also apply progressive enhancement information:

```tsx
// src/components/SignOut.tsx
import { actions } from 'astro:actions';

export function SignOut() {
  return (
    <form action={actions.signOut}>
      <button>Sign Out</button>
    </form>
  );
}
```

To learn more about Astro Actions, and to offer feedback on this experimental API, check out the [Astro Actions RFC](https://github.com/withastro/roadmap/blob/actions/proposals/0046-actions.md).

## Stabilized experimental features

> This release stabilize the CRSF protection (introduced in [Astro 4.6](/blog/astro-460/#experimental-support-for-csrf-protection)) and i18n domains support (introduced in [Astro 4.3](/blog/astro-430/#experimental-add-domain-support-for-i18n)) features. It is no longer necessary to enable the `experimental.csrfProtection` and `experimental.i18nDomains` options in your Astro config to use these features.

Reviewer: ```suggestion
This release stabilizes the CRSF protection (introduced in [Astro 4.6](/blog/astro-460/#experimental-support-for-csrf-protection)) and i18n domains support (introduced in [Astro 4.3](/blog/astro-430/#experimental-add-domain-support-for-i18n)) features. It is no longer necessary to enable the `experimental.csrfProtection` and `experimental.i18nDomains` options in your Astro config to use these features.
```
